[Console] Make sure signals registered by `SignalableCommandInterface` commands are defined in the PCNTL extension

alexandre-daubois opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.3
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | _NA_
| License       | MIT
| Doc PR        | No need I guess

Makes me think that adding a similar method to check a signal existence as a string in `SignalRegistry` could be leveraged in https://github.com/symfony/symfony/pull/49750

alexandre-daubois committed:
[Console] Make sure signals registered by `SignalableCommandInterface` commands are defined in the PCNTL extension

lyrixx commented:
Are we sure the signal list is finite?

alexandre-daubois commented:
@lyrixx I tried this snippet:

```php
<?php

function sig_handler($signo)
{
     switch ($signo) {
         default:
     }
}

echo "Installing signal handler";

// setup signal handlers
pcntl_signal(66,  "sig_handler");
```

And here is the result:

```
Installing signal handlerPHP Fatal error:  Uncaught ValueError: pcntl_signal(): Argument #1 ($signal) must be less than 65 in /home/alex/signals.php:13
Stack trace:
#0 /home/alex/signals.php(13): pcntl_signal()
#1 {main}
  thrown in /home/alex/signals.php on line 15
```

So we should at least check that the signal number is lower than 65.
Checking that the given signal is defined in the extension comes from this comment from Nicolas: https://github.com/symfony/symfony/pull/49750#discussion_r1144036134. What do you think?

lyrixx commented:
I think we should not enforce something that PHP allow. So :+1: with checking only if the number is lower than 65

lyrixx commented:
I think we should not enforce something that PHP allow. So :+1: with checking only if the number is lower than 65

lyrixx commented:
I think we should not enforce something that PHP allow. So :+1: with checking only if the number is lower than 65

> I think we should not enforce something that PHP allow. So +1 with checking only if the number is lower than 65

lyrixx commented:
Miss click, sorry! GitHub is really slow ATM 😬, I'll re-open as soon as GitHub allow me to!

nicolas-grekas commented on src/Symfony/Component/Console/SignalRegistry/SignalRegistry.php:
```diff
@@ -15,6 +15,8 @@ final class SignalRegistry
 {
     private array $signalHandlers = [];
 
+    private static ?array $availableSignals = null;
```
shouldn't be nullable

nicolas-grekas commented on src/Symfony/Component/Console/SignalRegistry/SignalRegistry.php:
```diff
@@ -42,6 +44,15 @@ public static function isSupported(): bool
+    public static function isValidSignal(int $signal): bool
+    {
+        if (null === static::$availableSignals) {
+            static::$availableSignals = get_defined_constants(true)['pcntl'];
```
`static::$availableSignals ??= get_defined_constants(true)['pcntl'] ?? [];`

nicolas-grekas commented on src/Symfony/Component/Console/Application.php:
```diff
@@ -1006,6 +1007,12 @@ protected function doRunCommand(Command $command, InputInterface $input, OutputI
 
+            foreach (array_merge($commandSignals, $this->signalsToDispatchEvent) as $signal) {
+                if (!SignalRegistry::isValidSignal($signal)) {
+                    throw new InvalidArgumentException(sprintf('Signal with code %d is not valid, make sure to use a signal defined in the `pcntl` extension by using a SIG* predefined constant.', $signal));
```
`Signal "%s" is not valid, make sure to use one of the "SIG*" constants as defined by the "pcntl" extension.`

alexandre-daubois replied:
Updated. I kept the back-ticks to be consistent with the exception thrown at line 1007 :slightly_smiling_face:

nicolas-grekas replied:
we don't use backticks in messages so this should be fixed the other way around

alexandre-daubois replied:
Okay, updated! I'm sending a PR on 5.4 to fix the other one (will search for other occurrences as well)

nicolas-grekas commented on src/Symfony/Component/Console/CHANGELOG.md:
```diff
@@ -7,6 +7,7 @@ CHANGELOG
  * Add support for choosing exit code while handling signal, or to not exit at all
  * Add `ProgressBar::setPlaceholderFormatter` to set a placeholder attached to a instance, instead of being global.
  * Add `ReStructuredTextDescriptor`
+ * Make sure signals registered by `SignalableCommandInterface` commands are defined in the PCNTL extension
```
should be removed, this is not a new feature but an improvement

alexandre-daubois commented:
I updated the `SignalRegistry` to check if the signal code is less than 32. Indeed, `pcntl_signal_get_handler()` in `SignalRegistry` if the signal code is above 31.

I also added a few tests for `SignalRegistry`.

nicolas-grekas commented:
> Fatal error:  Uncaught ValueError: pcntl_signal(): Argument #1 ($signal) must be less than 65 in /home/alex/signals.php:13

So, there is no need to add code to handle this, since the engine already has an appropriate behavior for invalid signals, isn't it?

lyrixx commented:
less code == less bug. I agree with @nicolas-grekas
